### ISSUE TITLE
Please give the option to update .db files periodically while FMD runs

### ISSUE THREAD

**Issue by pclkpg581** [fmd-project-team/FMD#879]
I leave FMD on continuously, and while the favorites.db-journal gets appended to favorites.db after it finishes checking for chapter updates, the other two do not. downloads.db and downloadedchapters.db do not get updated while FMD runs, they only get updated when you exit FMD.

This leads to a de-syncing of what is on the downloads list when FMD crashes/hangs. The favorites.db updates, but the download databases never updated, so the download queued chapters are lost. Countless chapter downloads get skipped due to this.

Would it be possible to have the other two .db files get updated after favorites.db does? That would at least ensure updated chapters do get saved in the download queue, and update/download progress is not lost. (Ideally, the databases would get updated whenever a new entry is added to any of the lists to minimize lost progress.)

Also, are .db-journal files readable? If so, would it be possible to append the usable parts of the .db-journal files to the .db files myself?

**Comment by Tmp341** [comment]
Just rename it's extension to ``.db``.

FMD is designed like that. I dunno if it is possible to change that now.

**Comment by dazedcat19** [comment]
There's used to be periodic database commit. but the disk write slowing down the system so much for some people. there's some people that keep so much download list (18K) which is unthinkable. fmd wasn't optimized to keep that much of list. always clean your completed download list. there is a lot of stuff need to be optimized. the download thread itself iterating every items which make the loop take long on large list. there is also a lot of this iteration on other methods. i did optimize some area. but there is still a lot need to be done.
i have a lot of ideas. like split the downloads at runtime between finished items and working items. so download thread only iterate working items. this also will need to change the interface between thread and UI to be virtual. a container to sync between multiple sub container and items. and push every changed item to the dbms. then commit it to disk periodically using timer.

in short it's not simple. since there's no experience pascal developer here beside myself, the development running slow. I have no time sitting for hours burning my head at the moment.

**Comment by pclkpg581** [comment]
Thanks for the replies, Tmp341 and dazedcat19.

Tmp341: The journal file seems to read correctly when I paste the contents to the end of the full .db. I tried reading it on its own after changing the extension to .db but it probably lacks some part of the full file; I used SQL Database Browser (https://sqlitebrowser.org/) for reading the files, and the journal fragment alone is interpreted as encrypted/corrupted.

dazedcat19: I did end up cleaning up my download list to minimize the possible hanging. Thanks for the insight on how the program handles the working data.

**Comment by spamtaztik** [comment]
I opened an issue ticket about this some time ago, but for some reason the discussion veered off to the manga list db from sites. Meh. Yes, this is what I was talking about. The download list is lost when FMD crashes/hangs/ PC froze. 

How about adding an option to close and reopens FMD every x hours? Even if FMD was in the middle of downloading something, after it's reopened it will resume downloading, no? I can't use windows task scheduler to run taskkill because it's the same as FMD crashing (won't save the download list).

Thoughts?

**Comment by dazedcat19** [comment]
I never experience any crash that corrupt my db. well I never run fmd for a long time. maybe half a day max. if there's a crash then there's a bug that causing it. instead of adding some messy code that does nothing to the bug. we should focus to find the bug instead.

**Comment by spamtaztik** [comment]
In my case it's not really FMD crashing, but more towards something wrong with the PC (run out of RAM or something) then it froze and has to be restarted. And the db is not corrupted, it's just the new downloads are gone. 

Say if I started FMD on April 1st and my PC froze on April 3rd. After restarting FMD, every single download from April 1st onward would be gone from the downloads page. FMD would load the old (April 1st) db files without first merging the changes in db-journal files.

**Comment by dazedcat19** [comment]
That doesn't make sense. there's nothing to commit when it started.

**Comment by bloodsign** [comment]
I guess I'm not the only one frequently having this issue. 
There were plenty of instances when FMD crashes where the downloaded list becomes desynced with the favorites. I usually have FMD running in the background and sometimes when I want to maximize it or send it to the foreground, it doesn't respond, which sometimes ends it up Not responding. 

The bigger problem is when the crash happens during download, all the download progress disappears while the favorites db marks it as downloaded. So re-running the "check for updates" skips those chapters since it is marked as downloaded. Which is a pain when i'm updating a huge number of chapters from my huge list of favorites.

**Comment by Tmp341** [comment]
FMD uses temp db files, so it will be overwritten when FMD successfully closed. But, in some cases like crushes, FMD leaves them there, and when it restarted, it doesn't see temp files. 

So, you can rename temp files before restarting FMD.

**Comment by bloodsign** [comment]
Just to add here. Shutting down windows without closing the app also causes this issue.

**Comment by spamtaztik** [comment]
@dazedcat19 the DB commit at the start is only used to commit any unsaved changes in case of FMD crashing. It should be (keyword is *should*, I don't know how the program is coded) just a simple call to the module that's used to commit changes, no? The commit when exiting FMD can be left as is.

@Tmp341 You said "renaming temp files"? How should I do this? Rename them to what? And after I rename them, then what? I don't see anything about this in the wiki... Instructions, pretty please?

**Comment by Tmp341** [comment]
@spamtaztik While update frozen/crashed, you should rename ``__xxxx.db`` to ``xxxx.db``

![sssss](https://user-images.githubusercontent.com/1683388/88843915-86107780-d1ea-11ea-80c0-18b9e195efbe.jpg)

I've marked, as you can see, there are 3 files. Rename higher size, in this case it is always ``__xxxx.db``, other files are waiting to be overwritten. Delete other 2 files.
